### PR TITLE
EDM-213: Added polarion case OCP-75523 A device can upgrade with

### PR DESCRIPTION
**Pull request by Samudelacruz77**
Definitive changes

### CODE REVIEW

**Review by hexfusion — COMMENTED** [review]
overall great progress one little nit then lgtm

File: test/e2e/agent/agent_updates_test.go [hunk 1]
```diff
@@ -75,6 +75,80 @@ var _ = Describe("VM Agent behavior during updates", func() {
...
+			// TODO(hexfusion): we were expecting this update status not to be unknown at this point
+			// related to: https://issues.redhat.com/browse/EDM-679
+			// Expect(device.Status.Updated.Status).ToNot(Equal(v1alpha1.DeviceUpdatedStatusType("Unknown")))
+			logrus.Infof("Device updated to new image %s 🎉", "flightctl-device:v4")
+
+			logrus.Info("Container with sleep infinity process is present but not running, as expected")
+			stdout, err := harness.VM.RunSSH([]string{"sudo", "podman", "ps"}, nil)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(stdout.String()).To(ContainSubstring("sleep infinity"))
+			logrus.Info("Pods are not running, as expected 👌")
```
[the remark this answers is not in the document]
Comment (Samudelacruz77) [reply test/e2e/agent/agent_updates_test.go:122]: Will change in next commit, thanks

File: test/e2e/agent/agent_updates_test.go [hunk 2]
```diff
@@ -75,6 +75,80 @@ var _ = Describe("VM Agent behavior during updates", func() {
...
+				func(device *v1alpha1.Device) bool {
+					return device.Status.Os.Image == newImageReference &&
+						conditionExists(device, "Updating", "False", "Updated")
+				}, "2m")
+
+			// TODO(hexfusion): we were expecting this update status not to be unknown at this point
+			// related to: https://issues.redhat.com/browse/EDM-679
+			// Expect(device.Status.Updated.Status).ToNot(Equal(v1alpha1.DeviceUpdatedStatusType("Unknown")))
+			logrus.Infof("Device updated to new image %s 🎉", "flightctl-device:v4")
+			logrus.Info("Pods with sleep infinity process are present but not running, as expected")
```
Comment (mangelajo) [inline test/e2e/agent/agent_updates_test.go:117]:
```suggestion
			logrus.Info("We expect containers with sleep infinity process to be present but not running")
```

Feel free to fix later. This is just logging, but it's better to be clear.

File: test/e2e/agent/agent_updates_test.go [hunk 3]
```diff
@@ -75,6 +75,80 @@ var _ = Describe("VM Agent behavior during updates", func() {
...
+			stdout, err := harness.VM.RunSSH([]string{"sudo", "podman", "ps"}, nil)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(stdout.String()).To(ContainSubstring("sleep infinity"))
+			logrus.Info("Containers are not running, as expected 👌")
+
```
Comment (mangelajo) [inline test/e2e/agent/agent_updates_test.go:122]: logrus.Infof("Going back to base image and checking that there is no application now")

File: test/e2e/agent/agent_updates_test.go [hunk 4]
```diff
@@ -75,6 +75,80 @@ var _ = Describe("VM Agent behavior during updates", func() {
...
+			harness.WaitForDeviceContents(deviceId, "status.Os.Image gets updated",
+				func(device *v1alpha1.Device) bool {
+					return device.Status.Os.Image == newImageReference &&
+						conditionExists(device, "Updating", "False", "Updated")
+				}, "2m")
+
+			logrus.Infof("Device updated to new image %s 🎉", "flightctl-device:base")
+			Expect(device.Spec.Applications).To(BeNil())
+			logrus.Info("Application demo_embedded_app is not present in new image 🌞")
```
Comment (mangelajo) [inline test/e2e/agent/agent_updates_test.go:150]: Should we also check that the specific application containers are not deployed anymore?

Comment (Samudelacruz77) [reply test/e2e/agent/agent_updates_test.go:150]: Yes we can do that, no issue in adding

File: test/e2e/agent/agent_updates_test.go (hunk 3 above)
Comment (Samudelacruz77) [reply test/e2e/agent/agent_updates_test.go:122]: Will change it

File: test/e2e/agent/agent_updates_test.go (hunk 2 above)
Comment (Samudelacruz77) [reply test/e2e/agent/agent_updates_test.go:117]: will change

Comment (hexfusion) [reply test/e2e/agent/agent_updates_test.go:117]: This came back maybe from rebase @Samudelacruz77

File: test/e2e/agent/agent_updates_test.go [hunk 5]
```diff
@@ -75,6 +75,87 @@ var _ = Describe("VM Agent behavior during updates", func() {
...
+
+			stdout1, err1 := harness.VM.RunSSH([]string{"sudo", "podman", "ps"}, nil)
```
Comment (hexfusion) [inline test/e2e/agent/agent_updates_test.go:153]: optional: this is just a style nit but its ok to reuse variables given proper scoping.
```suggestion
			stdout, err = harness.VM.RunSSH([]string{"sudo", "podman", "ps"}, nil)
```

File: test/e2e/agent/agent_updates_test.go (hunk 2 above)
Comment (Samudelacruz77) [reply test/e2e/agent/agent_updates_test.go:117]: Yes, will change again

File: test/e2e/agent/agent_updates_test.go (hunk 5 above)
Comment (Samudelacruz77) [reply test/e2e/agent/agent_updates_test.go:153]: The linter doesnt allow it

Comment (hexfusion) [reply test/e2e/agent/agent_updates_test.go:153]: note
`stdout, err =`  vs `stdout, err =:`

File: test/e2e/agent/agent_updates_test.go [hunk 6]
```diff
@@ -75,6 +75,86 @@ var _ = Describe("VM Agent behavior during updates", func() {
...
+				logrus.Infof("current image for %s is %s", deviceId, currentImage)
+				repo, _ := parseImageReference(currentImage)
+				newImageReference = repo + ":v4"
+				device.Spec.Os = &v1alpha1.DeviceOSSpec{Image: newImageReference}
+				logrus.Infof("updating %s to image %s", deviceId, device.Spec.Os.Image)
+			})
+
+			harness.WaitForDeviceContents(deviceId, "the device is upgrading to renderedVersion: 2",
+				func(device *v1alpha1.Device) bool {
+					return conditionExists(device, "Updating", "True", "Update")
```
[the remark this answers is not in the document]
Comment (hexfusion) [reply test/e2e/agent/agent_updates_test.go:99]: we might want to adjust conditionExists to take a status.UpdateState might make sense to create updateConditionExists to help enforce this pattern.  cc @mangelajo

**Pull request closed without merging**